### DISCUSSION TITLE
Feature request: ability to catch duplicate usernames

### DISCUSSION THREAD

**Discussion by sandalwoodbox** [fastapi-users/fastapi-users#598] [Ideas]
I'm building an application where, in addition to email, users can choose a unique username (which is how they will be identified to other users.) I have set up a database model that has a unique username field to enforce this constraint, and I can probably set up field-level validation via pydantic.

However, this won't handle race conditions where two separate requests try to add the same username at the same time and both get past that validation. In that case, it would be great to have a way to catch that unique constraint violation error and handle duplicate custom fields like `username` similarly to duplicate `email`.

**Comment by sandalwoodbox** [comment]
Update: handling this via pydantic seems to be discouraged https://github.com/samuelcolvin/pydantic/issues/1227#issuecomment-585723753 I think the argument they're making is reasonable.

**Comment by sandalwoodbox** [comment]
I was able to work around this with code like:

```
from fastapi_users import FastAPIUsers
from fastapi_users.models import BaseUserCreate
from fastapi_users.models import BaseUserDB
from fastapi_users.user import CreateUserProtocol


def get_create_user(
    old_create_user: CreateUserProtocol,
) -> CreateUserProtocol:
    async def create_user(
        user: BaseUserCreate,
        safe: bool = False,
        is_active: bool = None,
        is_verified: bool = None,
    ) -> BaseUserDB:
        user = cast(UserCreate, user)
        query = select([exists().where(users.c.username == user.username)])
        result = await database.fetch_one(query)
        if result and result.get("anon_1"):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="REGISTER_USERNAME_ALREADY_EXISTS",
            )
        return await old_create_user(user, safe, is_active, is_verified)

    return create_user


class MyAppUsers(FastAPIUsers):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.create_user = get_create_user(self.create_user)
```

This doesn't handle race conditions - but the implementation for email doesn't seem to handle race conditions either?

**Comment by sandalwoodbox** [comment]
Ideally it would also be nice to be able to catch both errors and display them, instead of users having to potentially fix a duplicate username and then a duplicate email in sequence.

**Comment by frankie567** [comment]
Hi there!

This is one the first design choice I had to make when I started this library. Currently, I'm not willing to support registration/authentication by username.

I see that you already have most of it working, though. Overriding `get_create_user` in your project is a good idea.

As for the race conditions, the unicity is guaranteed in database by a unique constraint. Even if two users pass the check at the same time (which is very unlikely), the database will enforce the e-mail to be unique and reject the second insertion. Admittedly, the second user would get a 500 error but again, the probability of such event happening is super-low. I don't think it's worth the effort to implement a protection (which would imply a distributed lock with Redis for example).

So, what you should do is to add a unique constraint on your username field. You can check in the code of your database adapter how we did it for e-mail: https://github.com/frankie567/fastapi-users/tree/master/fastapi_users/db
